File: src/main/java/com/projectswg/forwarder/resources/networking/data/ProtocolStack.java
package com.projectswg.forwarder.resources.networking.data;

import com.projectswg.forwarder.resources.networking.ClientServer;
import com.projectswg.forwarder.resources.networking.packets.Fragmented;
import com.projectswg.forwarder.resources.networking.packets.Packet;
import com.projectswg.forwarder.resources.networking.packets.SequencedPacket;
import me.joshlarson.jlcommon.log.Log;
import org.jetbrains.annotations.NotNull;

import java.net.InetSocketAddress;
import java.util.*;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.BiConsumer;

public class ProtocolStack {
	
	private final FragmentedProcessor fragmentedProcessor;
	private final InetSocketAddress source;
	private final BiConsumer<InetSocketAddress, byte[]> sender;
	private final ClientServer server;
	private final BlockingQueue<byte []> outboundRaw;
	private final ConnectionStream<SequencedPacket> inbound;
	private final ConnectionStream<SequencedOutbound> outbound;
	private final Packager packager;
	
	private InetSocketAddress pingSource;
	private int connectionId;
	
	public ProtocolStack(InetSocketAddress source, ClientServer server, BiConsumer<InetSocketAddress, byte[]> sender) {
		this.fragmentedProcessor = new FragmentedProcessor();
		this.source = source;
		this.sender = sender;
		this.server = server;
		this.outboundRaw = new LinkedBlockingQueue<>();
		this.inbound = new ConnectionStream<>();
		this.outbound = new ConnectionStream<>();
		this.packager = new Packager(outboundRaw, outbound, this);
		
		this.connectionId = 0;
	}
	
	public void send(Packet packet) {
		Log.t("Sending %s", packet);
		send(packet.encode().array());
	}
	
	public void send(byte [] data) {
		sender.accept(source, data);
	}
	
	public void sendPing(byte [] data) {
		InetSocketAddress pingSource = this.pingSource;
		if (pingSource != null)
			sender.accept(pingSource, data);
	}
	
	public InetSocketAddress getSource() {
		return source;
	}
	
	public ClientServer getServer() {
		return server;
	}
	
	public int getConnectionId() {
		return connectionId;
	}
	
	public short getRxSequence() {
		return inbound.getSequence();
	}
	
	public short getTxSequence() {
		return outbound.getSequence();
	}
	
	public void setPingSource(InetSocketAddress source) {
		this.pingSource = source;
	}
	
	public void setConnectionId(int connectionId) {
		this.connectionId = connectionId;
	}
	
	public SequencedStatus addIncoming(@NotNull SequencedPacket packet) {
		return inbound.addUnordered(packet);
	}
	
	public SequencedPacket getNextIncoming() {
		return inbound.poll();
	}
	
	public byte [] addFragmented(Fragmented frag) {
		return fragmentedProcessor.addFragmented(frag);
	}
	
	public void addOutbound(@NotNull byte [] data) {
		try {
			outboundRaw.put(data);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
	
	public short getFirstUnacknowledgedOutbound() {
		SequencedOutbound out = outbound.peek();
		if (out == null)
			return -1;
		return out.getSequence();
	}
	
	public void clearAcknowledgedOutbound(short sequence) {
		outbound.removeOrdered(sequence);
	}
	
	public synchronized void fillOutboundPackagedBuffer(int maxPackaged) {
		packager.handle(maxPackaged);
	}
	
	public synchronized int fillOutboundBuffer(SequencedOutbound [] buffer) {
		return outbound.fillBuffer(buffer);
	}
	
	@Override
	public String toString() {
		return String.format("ProtocolStack[server=%s, source=%s, connectionId=%d]", server, source, connectionId);
	}
	
	public static class ConnectionStream<T extends SequencedPacket> {
		
		private final PriorityQueue<T> sequenced;
		private final PriorityQueue<T> queued;
		
		private long sequence;
		
		public ConnectionStream() {
			this.sequenced = new PriorityQueue<>();
			this.queued = new PriorityQueue<>();
		}
		
		public short getSequence() {
			return (short) sequence;
		}
		
		public synchronized SequencedStatus addUnordered(@NotNull T packet) {
			if (SequencedPacket.compare(getSequence(), packet.getSequence()) > 0) {
				T peek = peek();
				return peek != null && peek.getSequence() == getSequence() ? SequencedStatus.READY : SequencedStatus.STALE;
			}
			
			if (packet.getSequence() == getSequence()) {
				sequenced.add(packet);
				sequence++;
				
				// Add queued OOO packets
				T queue;
				while ((queue = queued.peek()) != null && queue.getSequence() == getSequence()) {
					sequenced.add(queued.poll());
					sequence++;
				}
				
				return SequencedStatus.READY;
			} else {
				queued.add(packet);
				
				return SequencedStatus.OUT_OF_ORDER;
			}
		}
		
		public synchronized void addOrdered(@NotNull T packet) {
			packet.setSequence(getSequence());
			addUnordered(packet);
		}
		
		public synchronized void removeOrdered(short sequence) {
			T packet;
			List<Short> sequencesRemoved = new ArrayList<>();
			while ((packet = sequenced.peek()) != null && SequencedPacket.compare(sequence, packet.getSequence()) >= 0) {
				T removed = sequenced.poll();
				assert packet == removed;
				sequencesRemoved.add(packet.getSequence());
			}
			Log.t("Removed acknowledged: %s", sequencesRemoved);
		}
		
		public synchronized T peek() {
			return sequenced.peek();
		}
		
		public synchronized T poll() {
			return sequenced.poll();
		}
		
		public synchronized int fillBuffer(T [] buffer) {
			int n = 0;
			for (T packet : sequenced) {
				if (n >= buffer.length)
					break;
				buffer[n++] = packet;
			}
			return n;
		}
		
		public int size() {
			return sequenced.size();
		}
		
	}
	
	public enum SequencedStatus {
		READY,
		OUT_OF_ORDER,
		STALE
	}
	
}

File: src/main/java/com/projectswg/forwarder/resources/networking/data/ProtocolStack.kt
package com.projectswg.forwarder.resources.networking.data

import com.projectswg.forwarder.resources.networking.ClientServer
import com.projectswg.forwarder.resources.networking.packets.Fragmented
import com.projectswg.forwarder.resources.networking.packets.Packet
import com.projectswg.forwarder.resources.networking.packets.SequencedPacket
import me.joshlarson.jlcommon.log.Log
import java.net.InetSocketAddress
import java.util.*
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue

class ProtocolStack(val source: InetSocketAddress, val server: ClientServer, private val sendPacket: (InetSocketAddress, ByteArray) -> Unit) {
	
	private val fragmentedProcessor: FragmentedProcessor = FragmentedProcessor()
	private val outboundRaw: BlockingQueue<ByteArray> = LinkedBlockingQueue()
	private val inbound: ConnectionStream<SequencedPacket> = ConnectionStream()
	private val outbound: ConnectionStream<SequencedOutbound> = ConnectionStream()
	private val packager: Packager = Packager(outboundRaw, outbound)
	
	private var pingSource: InetSocketAddress? = null
	var connectionId: Int = 0
	
	val rxSourceSequence: Long
		get() = inbound.sourceSequence
	
	val rxSequence: Short
		get() = inbound.getSequence()
	
	val txSourceSequence: Long
		get() = outbound.sourceSequence
	
	val txSequence: Short
		get() = outbound.getSequence()
	
	val outboundPending: Int
		get() = outboundRaw.size
	
	fun send(packet: Packet) {
		Log.t("Sending %s", packet)
		send(packet.encode().array())
	}
	
	fun send(data: ByteArray) {
		sendPacket(source, data)
	}
	
	fun sendPing(data: ByteArray) {
		val pingSource = this.pingSource
		if (pingSource != null)
			sendPacket(pingSource, data)
	}
	
	fun getNextIncoming(): SequencedPacket? = inbound.poll()
	
	fun getFirstUnacknowledgedOutbound(): Short {
		val out = outbound.peek() ?: return -1
		return out.sequence
	}
	
	fun setPingSource(source: InetSocketAddress) {
		this.pingSource = source
	}
	
	fun addIncoming(packet: SequencedPacket): SequencedStatus {
		return inbound.addUnordered(packet)
	}
	
	fun addFragmented(frag: Fragmented): ByteArray? {
		return fragmentedProcessor.addFragmented(frag)
	}
	
	fun addOutbound(data: ByteArray) {
		try {
			outboundRaw.put(data)
		} catch (e: InterruptedException) {
			throw RuntimeException(e)
		}
		
	}
	
	fun clearAcknowledgedOutbound(sequence: Short) {
		outbound.removeOrdered(sequence)
	}
	
	@Synchronized
	fun fillOutboundPackagedBuffer(maxPackaged: Int) {
		packager.handle(maxPackaged)
	}
	
	@Synchronized
	fun fillOutboundBuffer(buffer: Array<SequencedOutbound?>): Int {
		return outbound.fillBuffer(buffer)
	}
	
	override fun toString(): String {
		return String.format("ProtocolStack[server=%s, source=%s, connectionId=%d]", server, source, connectionId)
	}
	
	class ConnectionStream<T : SequencedPacket> {
		
		private val sequenced: PriorityQueue<T> = PriorityQueue()
		private val queued: PriorityQueue<T> = PriorityQueue()
		
		var sourceSequence: Long = 0
			private set
		
		fun getSequence(): Short {
			return sourceSequence.toShort()
		}
		
		@Synchronized
		fun addUnordered(packet: T): SequencedStatus {
			if (SequencedPacket.compare(getSequence(), packet.sequence) > 0) {
				val peek = peek()
				return if (peek != null && peek.sequence == getSequence()) SequencedStatus.READY else SequencedStatus.STALE
			}
			
			if (packet.sequence == getSequence()) {
				sequenced.add(packet)
				sourceSequence++
				
				// Add queued OOO packets
				while (true) {
					val queue = queued.peek()
					if (queue == null || queue.sequence != getSequence())
						break
					sequenced.add(queued.poll())
					sourceSequence++
				}
				
				return SequencedStatus.READY
			} else {
				queued.add(packet)
				
				return SequencedStatus.OUT_OF_ORDER
			}
		}
		
		@Synchronized
		fun addOrdered(packet: T) {
			packet.sequence = getSequence()
			addUnordered(packet)
		}
		
		@Synchronized
		fun removeOrdered(sequence: Short) {
			val sequencesRemoved = ArrayList<Short>()
			while (true) {
				val packet = sequenced.peek()
				if (packet == null || SequencedPacket.compare(sequence, packet.sequence) < 0)
					break
				val removed = sequenced.poll()
				assert(packet === removed)
				sequencesRemoved.add(packet.sequence)
			}
			Log.t("Removed acknowledged: %s", sequencesRemoved)
		}
		
		@Synchronized
		fun peek(): T? {
			return sequenced.peek()
		}
		
		@Synchronized
		fun poll(): T? {
			return sequenced.poll()
		}
		
		@Synchronized
		fun fillBuffer(buffer: Array<T?>): Int {
			var n = 0
			for (packet in sequenced) {
				if (n >= buffer.size)
					break
				buffer[n++] = packet
			}
			return n
		}
		
		fun size(): Int {
			return sequenced.size
		}
		
	}
	
	enum class SequencedStatus {
		READY,
		OUT_OF_ORDER,
		STALE
	}
	
}
